File: 6-kyu/Tribonacci Sequence/javascript/Tribonacci Sequence.js
/* My Soultion */
function tribonacci(signature,n){
    let res = signature.slice(), sum;
    let len = signature.length;
    
    for (let i = 0; i < n - len; i++) {
      sum = 0;
      for (let j = i; j < i + len; j++) {
        sum += res[j];
      }
      res.push(sum)
    }
    return res.slice(0,n);
}

/* Best Pratices & Clever */
function tribonacci(signature,n){  
    for (var i = 0; i < n-3; i++) { // iterate n times
      signature.push(signature[i] + signature[i+1] + signature[i+2]); // add last 3 array items and push to trib
    }
    return signature.slice(0, n); //return trib - length of n
}

/* Another Soultions */
function tribonacci(signature,n) {
    const result = signature.slice(0, n);
    while (result.length < n) {
      result[result.length] = result.slice(-3).reduce((p,c) => p + c, 0);
    }
    return result;
}

/* ----------- */
function tribonacci(signature,n){
    while (signature.length < n) {
      signature.push(signature.slice(-3).reduce((a, b) => a + b));
    }
    return signature.slice(0, n);
}

File: 6-kyu/Tribonacci Sequence/python/Tribonacci Sequence.py
# My Solution 1
def gen_tribonacci(max, *args):
    a, b, c = args[0], args[1], args[2]
    for _ in range(max):
        yield a + b + c
        a, b, c = b, c, a + b + c

def tribonacci(signature, n):
    res = signature[:n]
    for v in gen_tribonacci(n - 3, *signature):
        res.append(v)
    return res

# My Solution 2
def tribonacci(signature, n):
    res = signature[:n]
    gen_tri = (sum(res[-3:]) for _ in range(n-3))
    for v in gen_tri:
        res.append(v)
    return res

# Best Practice & Clever
def tribonacci(signature, n):
    res = signature[:n]
    for i in range(n - 3): res.append(sum(res[-3:]))
    return res

# Best Practice & Clever 2
def tribonacci(signature,n):
    while len(signature) < n:
        signature.append(sum(signature[-3:]))
    
    return signature[:n]